Run only one pytest parametrization when generating optest

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108936

Richard, I'm curious to see what you think of this. I'm trying to use optest on the torchvision test suite, and after hacking up pytest support in https://github.com/pytorch/pytorch/pull/108929 I noticed that this was 5x'ing the test time... for no good reason.

* torchvision nms tests before optests: 60 passed, 4 skipped, 1206 deselected in 11.47s
* after optests: 300 passed, 20 skipped, 1206 deselected in 49.85s

It's no good reason because torchvision has parametrized the tests to get a spread of various random generation, but for checking schema or fake tensor, we don't actually need to test for different values.

This PR hacks up the codegen to replace pytest parametrize markers so that, instead of sampling many values, we sample only one value if you mark it with `opcheck_only_one`. There's a carveout for device parametrization, where we always run all those variants.

With this PR:

* reduced optests: 88 passed, 4 skipped, 1206 deselected in 13.89s

Companion torchvision PR which uses this at https://github.com/pytorch/vision/pull/7961

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
[RFC] Run only one pytest parametrization when generating optest

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "[RFC] Run only one pytest parametrization when generating optest"


Richard, I'm curious to see what you think of this. I'm trying to use optest on the torchvision test suite, and after hacking up pytest support in https://github.com/pytorch/pytorch/pull/108929 I noticed that this was 5x'ing the test time... for no good reason.

* torchvision nms tests before optests: 60 passed, 4 skipped, 1206 deselected in 11.47s
* after optests: 300 passed, 20 skipped, 1206 deselected in 49.85s

It's no good reason because torchvision has parametrized the tests to get a spread of various random generation, but for checking schema or fake tensor, we don't actually need to test for different values.

This PR hacks up the codegen to replace pytest parametrize markers so that, instead of sampling many values, by default we only sample the first value. There are more bells and whistles we could add; for example, I could add an extra custom pytest marker to let you say "no no, please run each parametrization in the optest". I could also not do this.

With this PR:

* reduced optests: 88 passed, 4 skipped, 1206 deselected in 13.89s

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -146,6 +146,21 @@ def new_method(*args, **kwargs):
                 result = method(*args, **kwargs)
             return result
 
+        if pytestmark := new_method.__dict__.get("pytestmark"):
+            import pytest
+
+            new_pytestmark = []
+            for mark in pytestmark:
+                if isinstance(mark, pytest.Mark) and mark.name == "parametrize":
+                    argnames, argvalues = mark.args
+                    assert not mark.kwargs, "NYI"
+                    new_pytestmark.append(
+                        pytest.mark.parametrize(argnames, (next(iter(argvalues)),))
+                    )
+                else:
+                    new_pytestmark.append(mark)
+            new_method.__dict__["pytestmark"] = new_pytestmark
+
```
I've thought about this before, especially because the internal tests also use some parametrization that make the tests go by slower. It kind of depends on how the parametrization is used:
- if it is just to generate different random values, then yes, one parametrization is sufficient.
- if it is to generate different options in your operator (e.g. BatchNorm with training=True/False), then it is important that we test all of them (BatchNorm mutates in one path but not the other)

Since generate_opcheck_tests is meant to be a "slow but runs everything check, because the user didn't want to write OpInfos (and we don't have an OpInfo V2 API)", I think I'd prefer the default to be "we run all parameterizations", but if the user understands what their test is doing and how it exercises their operator, they should be able to control the behavior to run just a single parametrization, perhaps by adding a decorator on the original test. WDYT?

ezyang replied:
This is fine by me. I'll need to figure out a mechanism to add extra annotations to parametrize though.

ezyang committed:
Update on "[RFC] Run only one pytest parametrization when generating optest"


Richard, I'm curious to see what you think of this. I'm trying to use optest on the torchvision test suite, and after hacking up pytest support in https://github.com/pytorch/pytorch/pull/108929 I noticed that this was 5x'ing the test time... for no good reason.

* torchvision nms tests before optests: 60 passed, 4 skipped, 1206 deselected in 11.47s
* after optests: 300 passed, 20 skipped, 1206 deselected in 49.85s

It's no good reason because torchvision has parametrized the tests to get a spread of various random generation, but for checking schema or fake tensor, we don't actually need to test for different values.

This PR hacks up the codegen to replace pytest parametrize markers so that, instead of sampling many values, by default we only sample the first value. There are more bells and whistles we could add; for example, I could add an extra custom pytest marker to let you say "no no, please run each parametrization in the optest". I could also not do this.

With this PR:

* reduced optests: 88 passed, 4 skipped, 1206 deselected in 13.89s

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Run only one pytest parametrization when generating optest"


Richard, I'm curious to see what you think of this. I'm trying to use optest on the torchvision test suite, and after hacking up pytest support in https://github.com/pytorch/pytorch/pull/108929 I noticed that this was 5x'ing the test time... for no good reason.

* torchvision nms tests before optests: 60 passed, 4 skipped, 1206 deselected in 11.47s
* after optests: 300 passed, 20 skipped, 1206 deselected in 49.85s

It's no good reason because torchvision has parametrized the tests to get a spread of various random generation, but for checking schema or fake tensor, we don't actually need to test for different values.

This PR hacks up the codegen to replace pytest parametrize markers so that, instead of sampling many values, by default we only sample the first value. There are more bells and whistles we could add; for example, I could add an extra custom pytest marker to let you say "no no, please run each parametrization in the optest". I could also not do this.

With this PR:

* reduced optests: 88 passed, 4 skipped, 1206 deselected in 13.89s

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Run only one pytest parametrization when generating optest"


Richard, I'm curious to see what you think of this. I'm trying to use optest on the torchvision test suite, and after hacking up pytest support in https://github.com/pytorch/pytorch/pull/108929 I noticed that this was 5x'ing the test time... for no good reason.

* torchvision nms tests before optests: 60 passed, 4 skipped, 1206 deselected in 11.47s
* after optests: 300 passed, 20 skipped, 1206 deselected in 49.85s

It's no good reason because torchvision has parametrized the tests to get a spread of various random generation, but for checking schema or fake tensor, we don't actually need to test for different values.

This PR hacks up the codegen to replace pytest parametrize markers so that, instead of sampling many values, we sample only one value if you mark it with `opcheck_only_one`. There's a carveout for device parametrization, where we always run all those variants.

With this PR:

* reduced optests: 88 passed, 4 skipped, 1206 deselected in 13.89s

Companion torchvision PR which uses this at https://github.com/pytorch/vision/pull/7961

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

zou3519 commented on torch/testing/_internal/optests/generate_tests.py:
```diff
@@ -147,6 +147,34 @@ def new_method(*args, **kwargs):
                 result = method(*args, **kwargs)
             return result
 
+        if pytestmark := new_method.__dict__.get("pytestmark"):
+            import pytest
```
Can you include a comment somewhere on how to actually mark a test with opcheck_only_one?
